### DISCUSSION TITLE
OSError: [Errno 98] EADDRINUSE

### DISCUSSION THREAD

**Discussion by andreasr81** [micropython/micropython#10438] [RP2040 / Pico]
Hi, i am a beginner, and done the code on https://projects.raspberrypi.org/en/projects/get-started-pico-w/5

at first all went fine, then i get this error and dont no why?!

```
error
```

Traceback (most recent call last):
  File "<stdin>", line 35, in <module>
  File "<stdin>", line 29, in open_socket
OSError: [Errno 98] EADDRINUSE

```
error
```

```
code
```

import network
import socket
from time import sleep
from picozero import pico_temp_sensor, pico_led
import machine

ssid = 'OpenWrt'
password = '**********'

def connect():
    #Connect to WLAN
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.connect(ssid, password)
    while wlan.isconnected() == False:
        print('Waiting for connection...')
        sleep(1)
    ip = wlan.ifconfig()[0]
    print(f'Connected on {ip}')
    return ip

def open_socket(ip):
    # Open a socket
    address = (ip, 80)
    connection = socket.socket()
    connection.bind(address)
    connection.listen(1)
    return connection

try:
    ip = connect()
    open_socket(ip)
except KeyboardInterrupt:
    machine.reset()

```
end
```

**Comment by davefes** [comment]
`connection.sockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)`
The keyword is REUSEADDR

**Reply by SlowFeather** [reply]
`AttributeError: 'socket' object has no attribute 'sockopt'`
He will trigger this error
`connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)`
This is available on Pico w

**Comment by andreasr81** [comment]
thanks for the answer, i find the solution in the meantime, it was connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1) nearly the same as your solution.

**Reply by HardwaterAK** [reply]
hello, I am having the same issue. could share were in the code you insert connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

**Reply by davefes** [reply]
This was in a server application:
```py
        if (sta_if.isconnected()):
            try:
                s = usocket.socket()

             #  allow for immediate re-use
                s.setsockopt(usocket.SOL_SOCKET, usocket.SO_REUSEADDR, 1)

                s.bind((config.host, config.port))
                s.settimeout(60)
                s.listen(0)

                print('listening for 60 seconds')
                ......             
```

**Reply by DGSArchos** [reply]
Hi ! I have this issue too. I didn't understand where you put those lines in the code.

**Comment by hwsacuity** [comment]
Answer ! of  this link https://stackoverflow.com/questions/6380057/address-already-in-use-error-when-binding-a-socket-in-python Worked but not sure if setting the port to 0  is a good idea in the address, but it   let me run a  test  adding  connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1) alone  was  still gave OSError: [Errno 98] EADDRINUSE  when   address = (ip, 80) is in the code.
def open_socket(ip):
    #Open a socket
    address = (ip, 0)
    connection = socket.socket()
    connection.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    connection.bind(address)
    connection.listen(1)
    print(connection)

**Reply by mattyrjackson** [reply]
Also had same issue and resolved it with port 0.
I'd be interested to know a bit more about why port 80 was temperamental yet port 0 worked first time.
